surefire-providers/surefire-testng/src/main/java/org/apache/maven/surefire/testng/conf/TestNG740Configurator.java

```diff
@@ -37,33 +35,45 @@
  *
  * @since 3.0.0-M6
  */
-public class TestNG740Configurator extends TestNG60Configurator
+public class TestNG740Configurator
+    extends TestNG60Configurator
 {
     @Override
     public void configure( XmlSuite suite, Map<String, String> options )
         throws TestSetFailedException
     {
-        String threadCountAsString = options.get( THREADCOUNT_PROP );
-        int threadCount = threadCountAsString == null ? 1 : parseInt( threadCountAsString );
-        suite.setThreadCount( threadCount );
-
-        String parallel = options.get( PARALLEL_PROP );
-        if ( parallel != null )
+        // if [threadcount] option in unspecified
+        if ( !options.containsKey( THREADCOUNT_PROP ) )
         {
-            if ( !"methods".equalsIgnoreCase( parallel ) && !"classes".equalsIgnoreCase( parallel ) )
-            {
-                throw new TestSetFailedException( "Unsupported TestNG parallel setting: "
-                    + parallel + " ( only METHODS or CLASSES supported )" );
-            }
-            Class enumClass = tryLoadClass( XmlSuite.class.getClassLoader(), "org.testng.xml.XmlSuite$ParallelMode" );
-            Enum<?> parallelEnum = Enum.valueOf( enumClass, parallel.toUpperCase() );
-            invokeSetter( suite, "setParallel", enumClass, parallelEnum );
+            // acquire default [threadcount] value to avoid superclass hardcoding to 1
+            options.put( THREADCOUNT_PROP, Integer.toString( suite.getThreadCount() ) );
         }
 
-        String dataProviderThreadCount = options.get( "dataproviderthreadcount" );
-        if ( dataProviderThreadCount != null )
+        // if [parallel] option is specified
+        if ( options.containsKey( PARALLEL_PROP ) )
         {
-            suite.setDataProviderThreadCount( Integer.parseInt( dataProviderThreadCount ) );
+            // try to load the [ParallelMode] enumeration
+            Class enumClass = tryLoadClass( XmlSuite.class.getClassLoader(), "org.testng.xml.XmlSuite$ParallelMode" );
+            // if enumeration loaded
+            if ( enumClass != null )
+            {
+                // extract [parallel] option
+                String parallel = options.remove( PARALLEL_PROP );
+                try
+                {
+                    // convert [parallel] option to corresponding constant
+                    Enum<?> parallelEnum = Enum.valueOf( enumClass, parallel.toUpperCase() );
+                    // set [XmlSuite] parallel mode to specified value
+                    invokeSetter( suite, "setParallel", enumClass, parallelEnum );
+                }
+                catch ( IllegalArgumentException e )
+                {
+                    throw new TestSetFailedException( "Unsupported TestNG parallel setting: " + parallel, e );
+                }
+            }
         }
+        
+        // invoke superclass handler
+        super.configure( suite, options );
```

Comment:
Check the inheritance because I think this is not ok either.

Reply:
The problem is that the super method `configure(...)` in `TestNGMapConfigurator` already calls the method `setParallel(String parallel)` in TestNG. This is what we want to avoid and that's why we made the fix in #339 .

Reply:
If you want to call the super method and these problems appear, it is a good practice to create abstract protected methods `setThreadCount()` and `setParallel()`, other too, in an interface - first in super type. This way the merge logic would be the same because it will be defined in some super type but every abstract protected method would implement it specifically.

Reply:
Look closer... The value of the [parallel] option is being **removed** from the map. I'm processing the option here, and the removal of the option avoids double-processing (with the resultant failure) in the superclass implementation.

Reply:
I could do as you suggest and extract the logic to `setThreadCount` and setParallel` methods. My current approach was intended to limit the scope of revisions to this single class. Would you prefer that I refactor the implementation? This would reduce the complexity a bit while expanding the scope of this PR.

Reply:
I just committed revisions that use the refactoring approach you recommended. This eliminates the need for careful management of the options map and makes the code more readable.